file.go

```diff
@@ -7,56 +7,89 @@ import (
-// files/directories as values that can be supplied to commands. For
-// directories, child files are accessed serially by calling `NextFile()`.
+// Node is a common interface for files, directories and other special files
+type Node interface {
+	io.Closer
+
+	// Size returns size of this file (if this file is a directory, total size of
+	// all files stored in the tree should be returned). Some implementations may
+	// choose not to implement this
+	Size() (int64, error)
+}
+
+// Node represents a regular Unix file
 type File interface {
-	// Files implement ReadCloser, but can only be read from or closed if
-	// they are not directories
-	io.ReadCloser
-
-	// FileName returns a filename associated with this file
-	FileName() string
-
-	// FullPath returns the full path used when adding with this file
-	FullPath() string
-
-	// IsDirectory returns true if the File is a directory (and therefore
-	// supports calling `NextFile`) and false if the File is a normal file
-	// (and therefor supports calling `Read` and `Close`)
-	IsDirectory() bool
-
-	// NextFile returns the next child file available (if the File is a
-	// directory). It will return (nil, io.EOF) if no more files are
-	// available. If the file is a regular file (not a directory), NextFile
-	// will return a non-nil error.
-	NextFile() (File, error)
+	Node
+
+	io.Reader
+	io.Seeker
 }
 
-type StatFile interface {
-	File
+// DirEntry exposes information about a directory entry
+type DirEntry interface {
+	// Name returns base name of this entry, which is the base name of referenced
+	// file
+	Name() string
 
-	Stat() os.FileInfo
+	// Node returns the file referenced by this DirEntry
+	Node() Node
 }
 
-type PeekFile interface {
-	SizeFile
+// DirIterator is a iterator over directory entries.
+// See Directory.Entries for more
+type DirIterator interface {
+	// DirEntry holds information about current directory entry.
+	// Note that after creating new iterator you MUST call Next() at least once
+	// before accessing these methods. Calling these methods without prior calls
+	// to Next() and after Next() returned false may result in undefined behavior
+	DirEntry
+
+	// Next advances iterator to the next file.
+	Next() bool
 
-	Peek(n int) File
-	Length() int
+	// Err may return an error after previous call to Next() returned `false`.
+	// If previous call to Next() returned `true`, Err() is guaranteed to
+	// return nil
+	Err() error
 }
 
-type SizeFile interface {
-	File
+// Directory is a special file which can link to any number of files.
+type Directory interface {
+	Node
 
-	Size() (int64, error)
+	// Entries returns a stateful iterator over directory entries.
+	//
+	// Example usage:
+	//
+	// it := dir.Entries()
+	// for it.Next() {
+	//   name := it.Name()
+	//   file := it.Node()
+	//   [...]
+	// }
+	// if it.Err() != nil {
+	//   return err
+	// }
+	//
+	// Note:
+	// - Some implementations of this functions may define some constraints in how
+	//   it can be used
```

Comment:
Can specify the *minimum* guarantees this function makes?

Really, I think it's fine saying you can only iterate over one directory at once. I'd expect users to use ls, etc. to *browse* through a directory tree and `get` to fetch an entire tree.

On the other hand, leaving an unbounded escape hatch on an interface is not going to end well. This function effectively defines no contract and can do whatever it wants.

Reply:
So this function is a bit tricky, but I think we can say that users can skip directories, and shouldn't traverse branches in parallel.

Skipping directories would be a special case for `MultipartFile`, but it's doable.

---

There may be a more type-safe way to do this, but whatever I think of makes code on the go-ipfs side far messier. For example things like allowing directory to return entries recursively, so:

Example tree:

```
/a: Dir
/a/f :file
/a/b: Dir
```

Calling iterator on `/` can get us following entries:
```
/a/f: File
/a/b: Dir
```

or just `/a: Dir`, depending if we return a 'recursive aware' directory. This has a nice property: directories have some control over which part of the tree needs to be accessed in sequence, but it still doesn't restrict parallel access to leaf nodes, so this isn't really helping us

Reply:
This interface is fine as-is, we just need to make sure that any interface we settle on is well defined.

Reply:
Pushed an improved doc for this